core/embed/rust/src/storagedevice/storage_device.rs

```diff
@@ -0,0 +1,690 @@
+const ROTATION: Field<u16> = Field::public(APP_DEVICE, 0x0F);
+const SLIP39_IDENTIFIER: Field<u16> = Field::private(APP_DEVICE, 0x10);
+const SLIP39_ITERATION_EXPONENT: Field<u8> = Field::private(APP_DEVICE, 0x11);
+const SD_SALT_AUTH_KEY: Field<Vec<u8, SD_SALT_AUTH_KEY_LEN_BYTES>> =
+    Field::public(APP_DEVICE, 0x12).exact();
+const INITIALIZED: Field<bool> = Field::public(APP_DEVICE, 0x13);
+const SAFETY_CHECK_LEVEL: Field<u8> = Field::private(APP_DEVICE, 0x14);
+const EXPERIMENTAL_FEATURES: Field<bool> = Field::private(APP_DEVICE, 0x15);
+
+const SAFETY_CHECK_LEVEL_STRICT: u8 = 0;
+const SAFETY_CHECK_LEVEL_PROMPT: u8 = 1;
+const DEFAULT_SAFETY_CHECK_LEVEL: u8 = SAFETY_CHECK_LEVEL_STRICT;
+const SAFETY_CHECK_LEVELS: [u8; 2] = [SAFETY_CHECK_LEVEL_STRICT, SAFETY_CHECK_LEVEL_PROMPT];
+
+// TODO: somehow determine the DEBUG_MODE value
+const DEBUG_MODE: bool = true;
+const AUTOLOCK_DELAY_DEFAULT: u32 = 10 * 60 * 1000; // 10 minutes
+const AUTOLOCK_DELAY_MINIMUM: u32 = if DEBUG_MODE {
+    10 * 1000 // 10 seconds
+} else {
+    60 * 1000 // 1 minute
+};
+// autolock intervals larger than AUTOLOCK_DELAY_MAXIMUM cause issues in the
+// scheduler
+const AUTOLOCK_DELAY_MAXIMUM: u32 = 0x2000_0000; // ~6 days
+
+const STORAGE_VERSION_CURRENT: u8 = 0x02;
+
+extern "C" fn storagedevice_is_version_stored() -> Obj {
+    let block = || Ok(VERSION.has().into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_version() -> Obj {
+    let block = || {
+        VERSION
+            .get()
+            .map_or(Ok(Obj::const_none()), |x| (&x as &[u8]).try_into())
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_version(version: Obj) -> Obj {
+    let block = || {
+        VERSION
+            .set(Buffer::try_from(version)?.try_into()?)?
+            .try_into()
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_is_initialized() -> Obj {
+    let block = || Ok(INITIALIZED.get().unwrap_or(false).into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_is_initialized(is_initialized: Obj) -> Obj {
+    let block = || INITIALIZED.set(is_initialized.try_into()?)?.try_into();
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_rotation() -> Obj {
+    let block = || Ok(ROTATION.get().unwrap_or(0).into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_rotation(rotation: Obj) -> Obj {
+    let block = || {
+        let rotation = rotation.try_into()?;
+
+        if ![0, 90, 180, 270].contains(&rotation) {
+            Err(Error::ValueError(cstr!("Not valid rotation")))
+        } else {
+            ROTATION.set(rotation)?.try_into()
+        }
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_label() -> Obj {
+    let block = || LABEL.get().map_or(Ok(Obj::const_none()), |x| x.try_into());
```

Comment:
In [ba44c83](https://github.com/trezor/trezor-firmware/pull/2214/commits/ba44c834a8c0c826cc6480f6e5c1c84a09fc12fa) I could not make it work for all the situations, as some of those are not `try_into`s, but only `into`s (u8, u16). Also in case of `Vec<u8>`, to return bytes, we need to use `(&v as &[u8]).try_into()` to return bytes, opposed to list of integers.

There is probably a lot to improve here